dev/bots/prepare_package.dart

```diff
@@ -39,153 +32,212 @@ class ProcessFailedException extends Error {
-    }
-    args.addAll(<String>[
+  Future<String> _createZipArchive(File output, Directory source) {
+    assert(Platform.isWindows);  // 7Zip is only available on Windows.
+    final List<String> commandLine = <String>[
+      '7za',
+      'a',
+      '-tzip',
+      '-mx=9',
       output.absolute.path,
       path.basename(source.absolute.path),
-    ]);
-
-    _runProcess(zip, args,
+    ];
+    return _runProcess(commandLine,
         workingDirectory: new Directory(path.dirname(source.absolute.path)));
   }
 
-  void createTarArchive(File output, Directory source) {
-    final List<String> args = <String>[
+  Future<String> _createTarArchive(File output, Directory source) {
+    return _runProcess(<String>[
+      'tar',
       'cJf',
       output.absolute.path,
       path.basename(source.absolute.path),
-    ];
-    _runProcess(tar, args,
-        workingDirectory: new Directory(path.dirname(source.absolute.path)));
+    ], workingDirectory: new Directory(path.dirname(source.absolute.path)));
+  }
+
+  Future<String> _runProcess(List<String> commandLine, {Directory workingDirectory}) async {
+    workingDirectory ??= _flutterRoot;
+    if (_subprocessOutput) {
+      stderr.write('Running "${commandLine.join(' ')}" in ${workingDirectory.path}.\n');
+    }
+    final List<int> output = <int>[];
+    final Completer<Null> stdoutComplete = new Completer<Null>();
+    final Completer<Null> stderrComplete = new Completer<Null>();
+    Process process;
+    Future<int> allComplete() async {
+      await stderrComplete.future;
+      await stdoutComplete.future;
+      return process.exitCode;
+    }
+
+    try {
+      process = await _processManager.start(
+        commandLine,
+        workingDirectory: workingDirectory.absolute.path,
+        environment: _environment,
+      );
+      process.stdout.listen(
+        (List<int> event) {
+          output.addAll(event);
+          if (_subprocessOutput) {
+            stdout.add(event);
+          }
+        },
+        onDone: () async => stdoutComplete.complete(),
+      );
+      if (_subprocessOutput) {
+        process.stderr.listen(
+          (List<int> event) {
+            stderr.add(event);
+          },
+          onDone: () async => stderrComplete.complete(),
+        );
+      } else {
+        stderrComplete.complete();
+      }
+    } on ProcessException catch (e) {
+      final String message = 'Running "${commandLine.join(' ')}" in ${workingDirectory.path} '
+          'failed with:\n${e.toString()}';
+      throw new ProcessFailedException(message, -1);
+    } catch (e) {
+      rethrow;
+    }
+
+    final int exitCode = await allComplete();
+    if (exitCode != 0) {
+      final String message = 'Running "${commandLine.join(' ')}" in ${workingDirectory.path} '
+          'failed with $exitCode.';
+      throw new ProcessFailedException(message, exitCode);
+    }
+    return new Future<String>.value(UTF8.decoder.convert(output).trim());
```

Comment:
The method is marked `async`, I believe you don't have to wrap return values in Futures here?

Reply:
Done.